File: src/waytogocoop/components/isotope_panel.py
"""Dash component for speculative isotope-effect controls."""

from __future__ import annotations

import dash_bootstrap_components as dbc
from dash import Input, clientside_callback, dcc, html

from waytogocoop.config import DEFAULT_ISOTOPE_EXPONENT, EXOTIC_MASS_RANGES
from waytogocoop.materials.isotopes import (
    ELEMENTS,
    humanize_half_life,
    natural_average_mass,
    nearest_isotope_info,
)

_C_MASS_VISIBILITY_JS = """
function(substrate, overlayer) {
    const isG = (v) => typeof v === 'string' && v.indexOf('Graphene') === 0;
    return (isG(substrate) || isG(overlayer)) ? {} : {display: 'none'};
}
"""

_EXOTIC_WARNING = (
    "HIGHLY SPECULATIVE — synthetic isotopes are radioactive (several half-lives are "
    "far too short to grow or measure a film), and masses between/beyond known "
    "isotopes are purely hypothetical. Outputs are what-if illustrations only."
)

# Elements with a mass slider in the panel, in layout order
_SLIDER_ELEMENTS = ("Fe", "Te", "Sb", "C")

_registered_prefixes: set[str] = set()
_registered_exotic_prefixes: set[str] = set()


def _register_c_mass_visibility(id_prefix: str) -> None:
    """Show the C-mass slider only for graphene-family materials, on first use."""
    if id_prefix in _registered_prefixes:
        return
    _registered_prefixes.add(id_prefix)

    from dash import Output

    clientside_callback(
        _C_MASS_VISIBILITY_JS,
        Output(f"{id_prefix}-c-mass-container", "style"),
        Input(f"{id_prefix}-substrate-dropdown", "value"),
        Input(f"{id_prefix}-overlayer-dropdown", "value"),
    )


def _isotope_marks(elem_data):
    """Slider marks for the stable isotopes of an element."""
    return {
        iso.mass_number: {"label": str(iso.mass_number)}
        for iso in elem_data.isotopes
    }


def _exotic_marks(elem_data):
    """Slider marks for stable plus synthetic isotopes (synthetic starred)."""
    marks = _isotope_marks(elem_data)
    for iso in elem_data.synthetic_isotopes:
        marks[iso.mass_number] = {"label": f"{iso.mass_number}{elem_data.symbol}*"}
    return marks


def _nearest_isotope_text(symbol: str, mass: float | None) -> str:
    """Readout string for the nearest known isotope to a slider value."""
    if mass is None:
        return ""
    info = nearest_isotope_info(symbol, float(mass))
    if info.kind == "hypothetical":
        return "hypothetical mass — no known isotope"
    if info.kind == "synthetic":
        return f"≈ {info.label}* (t½ ≈ {humanize_half_life(info.half_life_s)})"
    return f"≈ {info.label}"


def _register_exotic_mode_callbacks(id_prefix: str) -> None:
    """Wire the exotic-mode switch and nearest-isotope readouts, on first use."""
    if id_prefix in _registered_exotic_prefixes:
        return
    _registered_exotic_prefixes.add(id_prefix)

    from dash import Output, State, callback, no_update

    range_outputs = []
    value_states = []
    for sym in _SLIDER_ELEMENTS:
        slider_id = f"{id_prefix}-{sym.lower()}-mass"
        range_outputs.extend(
            [
                Output(slider_id, "min"),
                Output(slider_id, "max"),
                Output(slider_id, "marks"),
                Output(slider_id, "value"),
            ]
        )
        value_states.append(State(slider_id, "value"))

    @callback(
        Output(f"{id_prefix}-iso-exotic-warning", "is_open"),
        *range_outputs,
        Input(f"{id_prefix}-iso-exotic-mode", "value"),
        *value_states,
    )
    def _toggle_exotic_mode(exotic_on, *slider_values):
        exotic_on = bool(exotic_on)
        outputs = [exotic_on]
        for sym, value in zip(_SLIDER_ELEMENTS, slider_values, strict=True):
            elem = ELEMENTS[sym]
            if exotic_on:
                lo, hi = EXOTIC_MASS_RANGES[sym]
                outputs.extend([lo, hi, _exotic_marks(elem), no_update])
            else:
                # Mirror the Rust app's clamp_overrides_to_stable: a mass left
                # outside the stable range would otherwise keep driving the
                # physics while the exotic warning is hidden.
                lo = elem.isotopes[0].atomic_mass
                hi = elem.isotopes[-1].atomic_mass
                clamped = (
                    min(max(float(value), lo), hi) if value is not None else no_update
                )
                outputs.extend([lo, hi, _isotope_marks(elem), clamped])
        return tuple(outputs)

    @callback(
        *[Output(f"{id_prefix}-iso-{s.lower()}-nearest", "children")
          for s in _SLIDER_ELEMENTS],
        *[Input(f"{id_prefix}-{s.lower()}-mass", "value")
          for s in _SLIDER_ELEMENTS],
    )
    def _update_nearest_readouts(fe_mass, te_mass, sb_mass, c_mass):
        values = (fe_mass, te_mass, sb_mass, c_mass)
        return tuple(
            _nearest_isotope_text(sym, val)
            for sym, val in zip(_SLIDER_ELEMENTS, values, strict=True)
        )


def create_isotope_panel(id_prefix: str) -> dbc.Card:
    """Return a collapsible Card with isotope enrichment controls.

    Exposes per-element mass sliders for Fe, Te, Sb (Bi is monoisotopic),
    a BCS isotope exponent slider, and a comparison toggle. A C-mass slider
    is shown only when a graphene-family material is selected. An exotic-
    isotope switch (HIGHLY SPECULATIVE) widens the slider ranges to include
    synthetic isotopes and hypothetical masses; each slider gets a nearest-
    known-isotope readout.

    Parameters
    ----------
    id_prefix : str
        Prefix for Dash component IDs.
    """
    fe = ELEMENTS["Fe"]
    te = ELEMENTS["Te"]
    sb = ELEMENTS["Sb"]
    c = ELEMENTS["C"]

    fe_nat = natural_average_mass("Fe")
    te_nat = natural_average_mass("Te")
    sb_nat = natural_average_mass("Sb")
    c_nat = natural_average_mass("C")

    _register_c_mass_visibility(id_prefix)
    _register_exotic_mode_callbacks(id_prefix)

    return dbc.Card(
        dbc.CardBody(
            [
                html.H5("Isotope Effects (Speculative)", className="card-title"),
                dbc.Checklist(
                    id=f"{id_prefix}-isotope-enabled",
                    options=[{"label": "Enable", "value": "on"}],
                    value=[],
                    inline=True,
                    className="mb-2",
                ),
                html.Div(
                    id=f"{id_prefix}-isotope-controls",
                    children=[
                        dbc.Alert(
                            "SPECULATIVE - simplified models, qualitative results only.",
                            color="danger",
                            className="py-1 px-2 mb-2",
                            style={"fontSize": "0.8em"},
                        ),
                        # --- Exotic-isotope mode (HIGHLY SPECULATIVE) ---
                        dbc.Switch(
                            id=f"{id_prefix}-iso-exotic-mode",
                            label="Exotic isotopes (HIGHLY SPECULATIVE)",
                            value=False,
                            className="mb-1",
                        ),
                        dbc.Alert(
                            _EXOTIC_WARNING,
                            id=f"{id_prefix}-iso-exotic-warning",
                            color="danger",
                            is_open=False,
                            className="py-1 px-2 mb-2",
                            style={"fontSize": "0.8em"},
                        ),
                        # --- Fe mass ---
                        dbc.Label(f"Fe mass (amu) — natural: {fe_nat:.2f}"),
                        dcc.Slider(
                            id=f"{id_prefix}-fe-mass",
                            min=fe.isotopes[0].atomic_mass,
                            max=fe.isotopes[-1].atomic_mass,
                            step=0.01,
                            value=fe_nat,
                            marks=_isotope_marks(fe),
                            tooltip={"placement": "bottom", "always_visible": True},
                        ),
                        html.Div(
                            id=f"{id_prefix}-iso-fe-nearest",
                            className="small text-muted",
                        ),
                        html.Br(),
                        # --- Te mass ---
                        dbc.Label(f"Te mass (amu) — natural: {te_nat:.2f}"),
                        dcc.Slider(
                            id=f"{id_prefix}-te-mass",
                            min=te.isotopes[0].atomic_mass,
                            max=te.isotopes[-1].atomic_mass,
                            step=0.01,
                            value=te_nat,
                            marks=_isotope_marks(te),
                            tooltip={"placement": "bottom", "always_visible": True},
                        ),
                        html.Div(
                            id=f"{id_prefix}-iso-te-nearest",
                            className="small text-muted",
                        ),
                        html.Br(),
                        # --- Sb mass ---
                        dbc.Label(f"Sb mass (amu) — natural: {sb_nat:.2f}"),
                        dcc.Slider(
                            id=f"{id_prefix}-sb-mass",
                            min=sb.isotopes[0].atomic_mass,
                            max=sb.isotopes[-1].atomic_mass,
                            step=0.01,
                            value=sb_nat,
                            marks=_isotope_marks(sb),
                            tooltip={"placement": "bottom", "always_visible": True},
                        ),
                        html.Div(
                            id=f"{id_prefix}-iso-sb-nearest",
                            className="small text-muted",
                        ),
                        html.Br(),
                        # --- C mass (graphene-family materials only) ---
                        html.Div(
                            id=f"{id_prefix}-c-mass-container",
                            style={"display": "none"},
                            children=[
                                dbc.Label(f"C mass (amu) — natural: {c_nat:.2f}"),
                                dcc.Slider(
                                    id=f"{id_prefix}-c-mass",
                                    min=c.isotopes[0].atomic_mass,
                                    max=c.isotopes[-1].atomic_mass,
                                    step=0.001,
                                    value=c_nat,
                                    marks=_isotope_marks(c),
                                    tooltip={
                                        "placement": "bottom",
                                        "always_visible": True,
                                    },
                                ),
                                html.Div(
                                    id=f"{id_prefix}-iso-c-nearest",
                                    className="small text-muted",
                                ),
                                html.Br(),
                            ],
                        ),
                        # --- BCS isotope exponent ---
                        dbc.Label("BCS isotope exponent (α)"),
                        dcc.Slider(
                            id=f"{id_prefix}-isotope-alpha",
                            min=-0.5,
                            max=1.0,
                            step=0.01,
                            value=DEFAULT_ISOTOPE_EXPONENT,
                            marks={
                                -0.18: "-0.18 (inverse)",
                                0: "0",
                                0.4: "0.4 (consensus)",
                                0.5: "0.5 (BCS)",
                                0.81: "0.81 (FeSe)",
                            },
                            tooltip={"placement": "bottom", "always_visible": True},
                        ),
                        html.Br(),
                        # --- Comparison toggle ---
                        dbc.Checklist(
                            id=f"{id_prefix}-isotope-comparison",
                            options=[
                                {
                                    "label": "Show natural vs. enriched comparison",
                                    "value": "on",
                                }
                            ],
                            value=[],
                            inline=True,
                        ),
                        html.Hr(),
                        # --- Computed effects readout ---
                        html.Div(id=f"{id_prefix}-isotope-info"),
                    ],
                ),
            ]
        ),
        className="mb-3",
    )

File: crates/moire-desktop/src/ui/isotope_panel.rs
use egui::{Color32, Ui};
use moire_core::isotopes::{
    exotic_mass_range, humanize_half_life, natural_average_mass, nearest_isotope_info,
    ElementData, IsotopeKind, C, FE, SB, TE,
};

use crate::app::MoireApp;

const WARN_RED: Color32 = Color32::from_rgb(220, 50, 50);

/// Stable slider range: lightest..heaviest stable isotope (amu).
fn stable_mass_range(elem: &'static ElementData) -> (f64, f64) {
    let first = elem
        .isotopes
        .first()
        .expect("isotope table must not be empty");
    let last = elem
        .isotopes
        .last()
        .expect("isotope table must not be empty");
    (first.atomic_mass, last.atomic_mass)
}

/// Slider range for the current mode: stable isotopes, or the HIGHLY
/// SPECULATIVE hypothetical exploration range when exotic mode is on.
fn mass_range(elem: &'static ElementData, exotic: bool) -> (f64, f64) {
    if exotic {
        exotic_mass_range(elem.symbol).unwrap_or_else(|| stable_mass_range(elem))
    } else {
        stable_mass_range(elem)
    }
}

/// Nearest-isotope readout under a mass slider. Synthetic isotopes are
/// starred and show their humanized half-life.
fn nearest_isotope_readout(ui: &mut Ui, symbol: &str, mass: f64) {
    let info = nearest_isotope_info(symbol, mass);
    match info.kind {
        IsotopeKind::Stable => {
            ui.small(format!("≈ {} (stable)", info.label));
        }
        IsotopeKind::Synthetic => {
            let half_life = info
                .half_life_s
                .map(humanize_half_life)
                .unwrap_or_default();
            ui.small(format!("≈ {}* (t½ ≈ {})", info.label, half_life));
        }
        IsotopeKind::Hypothetical => {
            ui.small("hypothetical mass — no known isotope");
        }
    }
}

/// One element mass slider plus nearest-isotope readout.
/// Returns `true` if the override changed.
fn mass_slider(
    ui: &mut Ui,
    elem: &'static ElementData,
    override_mass: &mut Option<f64>,
    exotic: bool,
    step: f64,
    decimals: usize,
) -> bool {
    let nat = natural_average_mass(elem);
    let (min, max) = mass_range(elem, exotic);
    let mut mass = override_mass.unwrap_or(nat).clamp(min, max);

    ui.label(format!(
        "{} mass (amu) — nat: {:.prec$}",
        elem.symbol,
        nat,
        prec = decimals
    ));
    let changed = ui
        .add(egui::Slider::new(&mut mass, min..=max).step_by(step))
        .changed();
    if changed {
        *override_mass = Some(mass);
    }
    nearest_isotope_readout(ui, elem.symbol, mass);
    changed
}

/// Pull any out-of-range overrides back into the stable ranges (used when
/// exotic mode is switched off).
fn clamp_overrides_to_stable(app: &mut MoireApp) {
    let overrides = [
        (&FE, &mut app.fe_mass_override),
        (&TE, &mut app.te_mass_override),
        (&SB, &mut app.sb_mass_override),
        (&C, &mut app.c_mass_override),
    ];
    for (elem, override_mass) in overrides {
        if let Some(mass) = override_mass {
            let (min, max) = stable_mass_range(elem);
            *mass = mass.clamp(min, max);
        }
    }
}

/// Render isotope enrichment controls. Returns `true` if any value changed.
pub fn show_isotope_panel(ui: &mut Ui, app: &mut MoireApp) -> bool {
    let mut changed = false;

    ui.heading("Isotope Effects");
    ui.add_space(4.0);

    if ui
        .checkbox(&mut app.isotope_enabled, "Enable (Speculative)")
        .changed()
    {
        changed = true;
    }

    if !app.isotope_enabled {
        return changed;
    }

    ui.colored_label(WARN_RED, "SPECULATIVE - qualitative only");
    ui.add_space(4.0);

    // --- Exotic-isotope tier (HIGHLY SPECULATIVE) ---
    if ui
        .checkbox(&mut app.exotic_mode, "Exotic isotopes (HIGHLY SPECULATIVE)")
        .changed()
    {
        if !app.exotic_mode {
            clamp_overrides_to_stable(app);
        }
        changed = true;
    }
    if app.exotic_mode {
        ui.colored_label(
            WARN_RED,
            "Synthetic isotopes are radioactive — several half-lives are far \
             too short to grow or measure a film — and masses between/beyond \
             known isotopes are purely hypothetical. Outputs are what-if \
             illustrations only.",
        );
    }
    ui.add_space(4.0);

    let exotic = app.exotic_mode;

    // --- Fe / Te mass sliders ---
    changed |= mass_slider(ui, &FE, &mut app.fe_mass_override, exotic, 0.01, 2);
    changed |= mass_slider(ui, &TE, &mut app.te_mass_override, exotic, 0.01, 2);

    // --- Sb mass slider (only relevant for Sb-containing overlayers) ---
    let overlayer = app.overlayer_material();
    let substrate = app.substrate_material();
    let overlayer_has_sb = overlayer.formula == "Sb2Te3" || overlayer.formula == "Sb2Te";

    if overlayer_has_sb {
        changed |= mass_slider(ui, &SB, &mut app.sb_mass_override, exotic, 0.01, 2);
    }

    // --- C mass slider (only for graphene substrate or overlayer) ---
    let has_graphene =
        substrate.formula.starts_with("Graphene") || overlayer.formula.starts_with("Graphene");
    if has_graphene {
        changed |= mass_slider(ui, &C, &mut app.c_mass_override, exotic, 0.0001, 3);
    }

    ui.add_space(8.0);

    // --- BCS isotope exponent ---
    // Literature: α = 0.35–0.8 (iron chalcogenides); α = −0.18 (inverse, pnictides)
    ui.label("BCS isotope exponent (α):");
    if ui
        .add(egui::Slider::new(&mut app.isotope_alpha, -0.5..=1.0).step_by(0.01))
        .changed()
    {
        changed = true;
    }

    ui.add_space(4.0);

    // --- Reset button ---
    if ui.button("Reset to natural").clicked() {
        app.fe_mass_override = None;
        app.te_mass_override = None;
        app.sb_mass_override = None;
        app.c_mass_override = None;
        app.isotope_alpha = 0.4;
        app.exotic_mode = false;
        changed = true;
    }

    changed
}
